### PR TITLE
sanitize names when using them as tag value

### PR DESCRIPTION
**Pull request by replay**
related: https://github.com/graphite-project/graphite-web/pull/2458

### CODE REVIEW

**Comment by DanCech** [comment]
The biggest concern here is the impact that this will have on anyone who is currently using series with a `~` character in the name, since those series will be changed.  We could reduce that impact if we reduce the scope here to only stripping leading `~` signs, since technically that is the only place that it actually impacts the tagging system.

**Comment by replay** [comment]
Yeah, we could do that. Then i think I should update the docs accordingly, and update the validation function in raintank/schema as well to keep everything consistent

**Comment by replay** [comment]
Another question: What do we do if a metric name is exactly `~` and nothing else? 
We have a rule that says tag values must be `>=1` char long in https://github.com/graphite-project/graphite-web/blob/master/docs/tags.rst#carbon
We could just add it to the index with tag `name=""`, but that would only be queriable via regex because the query `name=` has another meaning. 
In the case of MT we could also consider that metric to have "invalid tags", and there's already a setting to define whether metrics with invalid tags should still get indexed via their other tags and the name or not, but in the case of carbon I'm not sure if there's a nice way to reject a metric name.

**Comment by DanCech** [comment]
The only real option available here would be to drop it as invalid afaik

**Comment by replay** [comment]
I updated the docs accordingly https://github.com/graphite-project/graphite-web/pull/2458/files

**Comment by replay** [comment]
Currently, if the resulting string is `""`, it would raise an Exception which gets caught here: https://github.com/graphite-project/carbon/blob/8e94e48f1d7605857b4e8ff3f5ac3646d55fd3f8/lib/carbon/cache.py#L42
So it would ignore the metric and log an error. I think that's probably what we want

**Comment by replay** [comment]
I've pushed a few more tests, so the test coverage report is happy

File: lib/carbon/util.py [hunk 1]
```diff
@@ -336,6 +336,24 @@ def __init__(classObj, name, bases, members):
...
+
+    for char in cls.prohibitedTagChars:
+      if char in tag:
+        raise Exception('Prohibited char "{char}" in tag "{tag}"'.format(char=char, tag=tag))
+
+    if ';' in value:
+      raise Exception('Character ";" is not allowed in tag value "{value}"'.format(value=value))
+
+    if value[0] == '~':
+      raise Exception('Tags are not allowed to start with "~" in tag "{tag}"'.format(tag=tag))
```
Comment (DanCech) [inline lib/carbon/util.py]:
```suggestion
      raise Exception('Tag values are not allowed to start with "~" in tag "{tag}"'.format(tag=tag))
```

File: lib/carbon/util.py [hunk 2]
```diff
@@ -336,6 +336,24 @@ def __init__(classObj, name, bases, members):
...
+
+  @classmethod
+  def validateTagAndValue(cls, tag, value):
+    """validate the given tag / value based on the specs in the documentation"""
+    if len(tag) == 0 or len(value) == 0:
+      raise Exception('Tag/Value may not be empty')
+
+    for char in cls.prohibitedTagChars:
+      if char in tag:
+        raise Exception('Prohibited char "{char}" in tag "{tag}"'.format(char=char, tag=tag))
```
Comment (DanCech) [inline lib/carbon/util.py]:
```suggestion
        raise Exception('Character "{char}" is not allowed in tag "{tag}"'.format(char=char, tag=tag))
```

File: lib/carbon/util.py [hunk 3]
```diff
@@ -336,6 +336,24 @@ def __init__(classObj, name, bases, members):
...
+
+    if ';' in value:
+      raise Exception('Character ";" is not allowed in tag value "{value}"'.format(value=value))
```
Comment (DanCech) [inline lib/carbon/util.py]:
```suggestion
      raise Exception('Character ";" is not allowed in value "{value}" of tag {tag}'.format(value=value, tag=tag))
```

File: lib/carbon/util.py [hunk 4]
```diff
@@ -336,6 +336,24 @@ def __init__(classObj, name, bases, members):
 
 
 class TaggedSeries(object):
+  prohibitedTagChars = ';!^='
+
+  @classmethod
+  def validateTagAndValue(cls, tag, value):
+    """validate the given tag / value based on the specs in the documentation"""
+    if len(tag) == 0 or len(value) == 0:
```
Comment (DanCech) [inline lib/carbon/util.py]:
```suggestion
    if len(tag) == 0:
```

File: lib/carbon/util.py [hunk 5]
```diff
@@ -336,6 +336,24 @@ def __init__(classObj, name, bases, members):
...
+      raise Exception('Tag/Value may not be empty')
```
Comment (DanCech) [inline lib/carbon/util.py]:
```suggestion
      raise Exception('Tag may not be empty')
    if len(value) == 0:
      raise Exception('Value for tag "{tag}" may not be empty'.format(tag=tag))
```

**Pull request merged**